crates/store/re_tf/src/transform_resolution_cache.rs

```diff
@@ -840,57 +1070,56 @@ impl TransformResolutionCache {
-                            .pinhole_projections
-                            .as_ref()
-                            .is_none_or(|pinhole_projections| pinhole_projections.is_empty())
-                    {
-                        per_timeline.per_source_frame_transforms.remove(source);
+                        // Remove source entry if it's empty.
+                        if source_transforms.frame_transforms.is_empty()
+                            && source_transforms
+                                .pose_transforms
+                                .as_ref()
+                                .is_none_or(|pose_transforms| pose_transforms.is_empty())
+                            && source_transforms
+                                .pinhole_projections
+                                .as_ref()
+                                .is_none_or(|pinhole_projections| pinhole_projections.is_empty())
+                        {
+                            per_timeline.per_source_frame_transforms.remove(source);
+                        }
                     }
                 }
 
-                // Remove empty source update mentions.
-                per_source_frame_updates.retain(|_, times| !times.is_empty());
+                // TODO(andreas): Remove empty source update mentions.
             }
 
-            // Remove timeline if it's empty.
+            // Remove entire timeline if it's empty.
             if per_timeline.per_source_frame_transforms.is_empty() {
                 self.per_timeline.remove(timeline);
             }
         }
     }
 }
 
-fn warn_about_missing_source_transforms_for_update_on_entity(
+fn debug_panic_missing_source_transforms_for_update_on_entity(
     entity_path: &EntityPath,
     source: TransformFrameIdHash,
 ) {
-    // There was no actual transform changes for this source frame after all.
-    re_log::error_once!(
-        "Internally inconsistent state: entity {:?} had updates for source frame {:?} but no transforms for that source frame were found. Please report this as a bug.",
-        entity_path,
-        source,
-    );
+    // There was no actual transform change for this source frame after all.
+    if cfg!(debug_assertions) {
+        panic!(
+            "DEBUG ASSERTION: Internally inconsistent state: entity {entity_path:?} had updates for source frame {source:?} but no transforms for that source frame were found. Please report this as a bug."
+        );
+    }
 }
 
 /// Queries all components that are part of pose transforms, returning the transform from child to parent.
 ///
-/// If any of the components yields an invalid transform, returns a `Affine3A::ZERO`.
-/// (this effectively disconnects a subtree from the transform hierarchy!)
+/// If any of the components yields an invalid transform, returns `None`.
 // TODO(#3849): There's no way to discover invalid transforms right now (they can be intentional but often aren't).
 fn query_and_resolve_tree_transform_at_entity(
```

Comment:
I feel like this is a bit scary, similar to #8952 people could be depending on `Affine3A::ZERO` for invalid transforms.

I think this is the right thing to do, but might be worth to make a note of this somewhere in the changelog.